### DISCUSSION TITLE
Finding a workaround to common properties shared between slices.

### DISCUSSION THREAD

**Discussion by KamaniBhavin** [pmndrs/zustand#2518] [General]
I have two distinct slices, each with identical property names. This naming convention enhances code readability. However, upon creating a store, they merge into a unified global object, potentially overwriting properties from different slices. Is there a workaround for this issue? While I recognize it may not be directly related to Zustand, any assistance in addressing this matter would be greatly appreciated.

```ts

export interface A {
 initialize: () => void;
}

export const a: StateCreator<A, [["zustand/immer", never]]> = (set) => ({
  initialize: () => {
    // something specific to slice A
  },
});

export interface B {
  initialize: () => void;
}

export const b: StateCreator<B, [["zustand/immer", never]]> = (set) => ({
  initialize: () => {
    // something specific to slice B
  }
});

const store = create()(
  immer((...args) => ({ //<-------- we overwrite the initialise in A with B here.
    ...a(...args),
    ...b(...args),
    })
  ));
```

But having a property with the same name in both slices will cause a conflict. 
I would like to avoid this and have a way to define the slices in a way that they don't conflict with each other.
Is there a way to achieve this?

**Comment by dbritto-dev** [comment]
@KamaniBhavin you could try with zustand-lens

**Reply by KamaniBhavin** [reply]
@dbritto-dev  I'm avoiding adding extra dependencies or complexity, which is why I'm not utilizing `zustand-lens`. However, I've managed to find a workaround by creating a completely separate namespace.

```ts
import { create, StateCreator } from "zustand";
import { immer } from "zustand/middleware/immer";

export interface A {
  a: { // <--- different namespace for each slice state to avoid properties overriding when merging objects 
    initialize: () => void;
  };
}

export const a: StateCreator<A, [["zustand/immer", never]]> = (set) => ({
  a: {
    initialize: () => {
      // something specific to slice A
    },
  }
});

export interface B {
  b: { // <--- different namespace for each slice state to avoid properties overriding when merging objects 
    initialize: () => void;
  };
}

export const b: StateCreator<B, [["zustand/immer", never]]> = (set) => ({
  b: {
    initialize: () => {
      // something specific to slice B
    }
  }
});

const store = create()(
  immer((...args) => ({
    ...a(...args),
    ...b(...args),
    })
  ));
  ```
  
  Would it have any adverse effects?

**Reply by dbritto-dev** [reply]
@KamaniBhavin that's fine, just be careful with your types:

```ts
StateCreator<State, [['zustand/immer', never]], [], Slice>
```

**Reply by Sebastian-Nielsen** [reply]
It kind of sucks because:

For something that looked like this before (so simple)
```ts
export const myNamespace: StateCreator<A, [["zustand/immer", never]]> = (set) => ({
         setSomething: (b: boolean) => set({ something : b }),
  }
});
```
Now we have to do this: (so ugly)
```ts
export const myNamespace: StateCreator<A, [["zustand/immer", never]]> = (set) => ({
  myNamespace: {
         setSomething: (b: boolean) => set(state => ({ myNamespace: {...state.myNamespace, something: b} })),
    },
  }
});
```

So, in short, it sucks because I have to add this part `state => ({myNamespace: {...state.myNamespace, }` to all my setters.

**Reply by dbritto-dev** [reply]
@Sebastian-Nielsen also, you need to setup the type returned by the creator function

**Reply by Sebastian-Nielsen** [reply]
@dbritto-dev as a maintainer, would it be possible to have zustand automatically do all this namespacing here behind the scenes whenever you create a slice? 

This namespacing issue seems like something zustand should be able to handle does it not?

**Reply by dbritto-dev** [reply]
@KamaniBhavin the issue you are facing is not related to zustand and should be handled by the user

**Reply by dbritto-dev** [reply]
@KamaniBhavin for instance on redux you have reducers but if for some reason you use the same name on a different reducer one would overwrite the other one

**Reply by dbritto-dev** [reply]
@Sebastian-Nielsen would be better to spend some time trying zustand-lens

**Reply by Sebastian-Nielsen** [reply]
@KamaniBhavin The immer solution is good but not great because of the need to repeatedly reference `myNamespace`, which reduces readability. I am curious, how do you write your actions (setter functions)? Are you repeatedly referencing the namespace in question like this:

```ts
export const myNamespace: StateCreator<A, [["zustand/immer", never]]> = (set) => ({
  myNamespace: {
         setSomething: (b: boolean) => set(state => ({ myNamespace: {...state.myNamespace, something: b} })),
    },
  }
});
```
or 
```ts
export const myNamespace: StateCreator<A, [["zustand/immer", never]]> = (set) => ({
  myNamespace: {
         setSomething: (b: boolean) => set(state => { 
             state.myNamespace.something = b;
         }),
    },
  }
});
```

**Comment by dai-shi** [comment]
> This namespacing issue seems like something zustand should be able to handle does it not?

The short answer is no.

I wouldn't try to convince anyone, but here's some notes for anyone interested.
- Zustand philosophy is "unpoinionated"
- Namespacing is an important feature, but I decided it's out of the scope of Zustand core (this repo).
  - Quite a few use cases can be done without namespacing.
  - Namespacing requirements vary from person to person.
  - It feels like there are some API options.
- We currently have zustand-lens and zustand-slices. They shouldn't cover all use cases. There should be more libraries.
- Docs should clarify about this.

**Reply by Sebastian-Nielsen** [reply]
At the very least consider warning people in the docs and also maybe suggest what to do to avoid the risk:

When following the recommended "one store, multiple slices" pattern, you are constantly suceptible to the risk of introducing a bug if you ever happen to accidentally use the same attribute-name in two different slices at once. Unless you fix the namespacing issue yourself by ... (and then some suggestions maybe)

**Reply by dai-shi** [reply]
TBH, I’ve never liked the current slice pattern. I warned about name collisions and didn’t personally recommend the pattern. But the community liked it (even with name collisions), so we added it to the docs.

Now, my dislike of the current slice pattern has driven me to create zustand-slices. And @dbritto-dev is working on new docs, so you are right on time.

**Reply by dbritto-dev** [reply]
@Sebastian-Nielsen suggestions are welcome -> https://gist.github.com/dbritto-dev/7fc6abc8afb33e5f98df35e6e0aa7657
